lib/amo-client.js

```diff
@@ -37,25 +49,98 @@ function Client(conf) {
  * @return {Promise}
  */
 Client.prototype.sign = function(conf) {
+  var self = this;
   conf = conf || {};
 
-  var apiUrl = '/addons/' + encodeURIComponent(conf.guid);
-  apiUrl += '/versions/' + encodeURIComponent(conf.version) + '/';
+  var addonUrl = '/addons/' + encodeURIComponent(conf.guid);
+  addonUrl += '/versions/' + encodeURIComponent(conf.version) + '/';
 
   return this.put({
-    url: apiUrl,
+    url: addonUrl,
     formData: {
       upload: this._fs.createReadStream(conf.xpiPath),
     },
+  }, {
+    throwOnBadResponse: false,
   }).then(function(responseResult) {
     var httpResponse = responseResult[0];
     var response = responseResult[1];
 
-    console.log('Your add-on file is being processed');
-    console.log('Job ID:', response.id)
-    console.log('Not yet implemented: polling until completion, downloading the signed XPI');
+    var handledStatuses = [200, 409];
+    if (handledStatuses.indexOf(httpResponse.statusCode) === -1) {
+      throw new Error(
+        'Received bad response from the server while signing; ' +
+        'status: ' + httpResponse.statusCode + '; ' +
+        'response: ' +
+        (typeof response === 'string' ? response.substring(0, 100) + '...' : ''));
+    }
+
+    // TODO: maybe return early for 409s (already validated).
+    // Right now they are handled like 200s which seems wrong.
+
+    return self.waitForSignedAddon(addonUrl);
+  });
+};
+
+
+/**
+ * Poll a status URL, waiting for the queued add-on to be signed.
+ *
+ * @param {String} URL to GET for add-on status.
+ * @return {Promise}
+ */
+Client.prototype.waitForSignedAddon = function(statusUrl, opt) {
+  opt = _.assign({
+    clearTimeout: clearTimeout,
+    setAbortTimeout: setTimeout,
+    setStatusCheckTimeout: setTimeout,
+    abortAfter: 45000,
+  }, opt);
+  var self = this;
+
+  return when.promise(function(resolve, reject) {
+    self._validateProgress.animate();
+    var statusCheckTimeout;
+    var nextStatusCheck;
+
+    function checkSignedStatus() {
+      self.get({url: statusUrl}).then(function(result) {
+        var data = result[1];
+        if (data.valid && data.download_url) {
```

Comment:
This looks like it will keep checking until it passes validation which I think would be forever if it fails validation.

You should be able to check `data.processed` to know when automated validation has been completed. Then `data.valid` should let you know if it passed automated validation. Once it is valid it will get reviewed (which is automatic but still done asynchronously for unlisted) and then `data.reviewed` should become `tru

 If it has passed validation then `data.active` should tell you if the version has been created and signed.